src/System.Private.CoreLib/shared/System/Number.Grisu3.cs

```diff
@@ -311,410 +308,621 @@ internal static unsafe class Grisu3
                 0xAF87023B9BF0EE6B,
             };
 
-            public static bool Run(double value, int precision, ref NumberBuffer number)
+            private static readonly uint[] s_SmallPowersOfTen = new uint[]
             {
-                // ========================================================================================================================================
-                // This implementation is based on the paper: http://www.cs.tufts.edu/~nr/cs257/archive/florian-loitsch/printf.pdf
-                // You must read this paper to fully understand the code.
-                //
-                // Deviation: Instead of generating shortest digits, we generate the digits according to the input count.
-                // Therefore, we do not need m+ and m- which are used to determine the exact range of values.
-                // ======================================================================================================================================== 
-                //
-                // Overview:
-                //
-                // The idea of Grisu3 is to leverage additional bits and cached power of ten to produce the digits.
-                // We need to create a handmade floating point data structure DiyFp to extend the bits of double.
-                // We also need to cache the powers of ten for digits generation. By choosing the correct index of powers
-                // we need to start with, we can eliminate the expensive big num divide operation.
-                //
-                // Grisu3 is imprecision for some numbers. Fortunately, the algorithm itself can determine that and give us
-                // a success/fail flag. We may fall back to other algorithms (For instance, Dragon4) if it fails.
-                //
-                // w: the normalized DiyFp from the input value.
-                // mk: The index of the cached powers.
-                // cmk: The cached power.
-                // D: Product: w * cmk.
-                // kappa: A factor used for generating digits. See step 5 of the Grisu3 procedure in the paper.
-
-                // Handle sign bit.
-                if (double.IsNegative(value))
+                1,          // 10^0
+                10,         // 10^1
+                100,        // 10^2
+                1000,       // 10^3
+                10000,      // 10^4
+                100000,     // 10^5
+                1000000,    // 10^6
+                10000000,   // 10^7
+                100000000,  // 10^8
+                1000000000, // 10^9
+            };
+
+            public static bool RunDouble(double value, int requestedDigits, ref NumberBuffer number)
+            {
+                double v = double.IsNegative(value) ? -value : value;
+
+                Debug.Assert(v > 0);
+                Debug.Assert(double.IsFinite(v));
```

Comment:
What's guaranteeing this?

Reply:
The links are outdated now, but I was able to follow where the checks are happening. Thanks :)